Implement SM2DH_MLKEM768_HYBRID tls key exchange with kyber-ref.

pr000000f opened the pull request:
为 TLS 1.3 添加 SM2DH_MLKEM768_HYBRID 混合抗量子密钥交换协议，其中使用了 MLKEM 算法设计团队提供的参考实现。

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/Tongsuo-Project/Tongsuo?pullRequest=733) <br/>All committers have signed the CLA.

pr000000f committed:
Implement SM2DH_MLKEM768_HYBRID tls key exchange with kyber-ref.

vincehong reviewed (commented):
Great job, thanks! Below are a few suggestions to enhance the pull request:

1. To ensure compliance with the Apache License 2.0 of the Kyber reference code, please copy the LICENSE file from the Kyber repository and place it in the crypto/kyber/ directory (e.g., crypto/kyber/LICENSE). 
2. Please include a test case for the TLS handshake, potentially under test/ssl-tests/.
3. The codebase mixes naming conventions, such as SM2DH_MLKEM768_HYBRID (uppercase) and sm2dh_mlkem768_hybrid (lowercase). If it's not necessary, consider use a consistent naming.

vincehong commented on providers/implementations/keymgmt/sm2dh_mlkem768_hybrid_kmgmt.c:
```diff
@@ -0,0 +1,353 @@
+        // secret keys.
+        // (TODO: Find a better approach.)
+        if(buf_len == SM2_DH_MLKEM_768_HYBRID_PK_SIZE) {
+            memcpy(hybrid_key->pk, buf, SM2_DH_MLKEM_768_HYBRID_PK_SIZE);
```
Should we check that hybrid_key->pk and hybrid_key->ct are non-NULL ?

vincehong commented on providers/implementations/keymgmt/sm2dh_mlkem768_hybrid_kmgmt.c:
```diff
@@ -0,0 +1,353 @@
+        // since the client and server use the same function to parse the encoded public key
+        // to s_key and c_key, where both s_key and c_key are empty EVP_PKEY objects without
+        // secret keys.
+        // (TODO: Find a better approach.)
```
Is it possible to design an attack if I know the other party are using buf_len to distinguish public key and ciphertext?

pr000000f committed:
Add tls test and kyber's license, standardize naming case, and use a flag to decode ct and pk.

pr000000f commented:
> Great job, thanks! Below are a few suggestions to enhance the pull request:
> 
> 1. To ensure compliance with the Apache License 2.0 of the Kyber reference code, please copy the LICENSE file from the Kyber repository and place it in the crypto/kyber/ directory (e.g., crypto/kyber/LICENSE).
> 2. Please include a test case for the TLS handshake, potentially under test/ssl-tests/.
> 3. The codebase mixes naming conventions, such as SM2DH_MLKEM768_HYBRID (uppercase) and sm2dh_mlkem768_hybrid (lowercase). If it's not necessary, consider use a consistent naming.

Thank you!
In recent commits, we introduced the following changes:
1. Added kyber-ref's license file at crypto/kyber/LICENSE, originally from https://github.com/pq-crystals/kyber/blob/main/LICENSE.
2. Added a TLS handshake test for SM2DH and SM2DH-MLKEM768-HYBRID at test/ssl-tests/32-pqc-groups.cnf
3. Used a consistent naming of uppercase in the codebase.
4. Used a flag to distinguish the cases of public-key decoding and ciphertext decoding.
5. Fixed a few bugs causing CI failures.
